### ISSUE TITLE
Upgrade from 0.34.0 to 0.36.0 causes consumer to stop consuming but keep running

### ISSUE THREAD

**Issue by zhrebicek** [fede1024/rust-rdkafka#638]
As the title says, after upgrade from `0.34.0` to `0.36.0` we started getting Kafka consumers that would stop consuming messages, start leaking memory and do not leave the consumer group.
* I do not have clear reproducer as it happened to use across multiple production environments in random fashion.
  * And for example in our staging it never happened.
* It happened more often to environments where Kafka was under higher load.
Maybe it has to do with change to use event interface, but that is just wild guess.
* In the end I tried to set low [max.poll.interval.ms](http://max.poll.interval.ms/) but found out it happens again, which seems like it is still sending heartbeats, but for whatever reason stops pulling records, CPU drops down, memory increases and the consumer group does not re-balance.
* After revert to use `0.34.0`

<img width="773" alt="Screenshot 2023-12-12 at 14 40 59" src="https://github.com/fede1024/rust-rdkafka/assets/10925068/5acad8f5-5f4f-4ccd-8c42-2e16fce4f69b">

<img width="779" alt="Screenshot 2023-12-12 at 14 40 33" src="https://github.com/fede1024/rust-rdkafka/assets/10925068/07494f6e-5254-464c-a5b3-5bca94c4df21">

Disclaimer: Running on the `0.34.0` for half day without it reproducing and it happened many times with `0.36.0`. Would report back if it still happens on `0.34.0` and the origin is elsewhere.

**Comment by scanterog** [comment]
Thanks for reporting it, we'll investigate further. `0.35.0` does not contain the event api change if you wanna give it a try.

**Comment by neoeinstein** [comment]
Confirming that I too encountered a similar issue, which resulted in significant increases in idle time being reported across the board, and some instances of what appeared to be an executor stall (almost like if some blocking calls were on the async path). I happened to be using a producer for this, with no consumers on my particular service, though I do use a long publish batching window of 1 second.

Downgrading from `0.36.0` to `0.35.0` did indeed resolve the issue.

**Comment by davidblewett** [comment]
@zhrebicek @neoeinstein The event API does allow some transient errors to be propagated that the callback API did not.  @scanterog brought this up to upstream here: https://github.com/confluentinc/librdkafka/issues/4493 . Is it possible that your code might have not handled these explicitly, leading to either the consumer or producer becoming idle?

This may manifest with `KafkaError (Message consumption error: BrokerTransportFailure (Local: Broker transport failure))` or `KafkaError (Message consumption error: BrokerTransportFailure (Local: Broker transport failure))`.

**Comment by scanterog** [comment]
We tested internally today the `BaseProducer` and `BaseConsumer` and we couldn't reproduce the issue. We'll try out tomorrow the `FutureProducer` and the `StreamConsumer`. Please provide the properties you're setting and any relevant logs. Are you also using the statically linked librdkafka or linking it dynamically?

**Comment by neoeinstein** [comment]
On our side, we are using `FutureProducer`. We also record the amount of time that it takes to confirm the publish, but move that off the main execution path by `tokio::spawn`ing a future wrapping that `.send`. In general, each instance of this service is publishing about 100 messages per second into rdkafka, with an average of about 40 such spawned tasks pending concurrently (within a 1 second batching window).

**Comment by zhrebicek** [comment]
We use `rdkafka` as is `rdkafka = "0.34.0"`, running on `debian-bookworm-slim`.
We use `StreamConsumer`, and settings:
```yaml
client.id: <id>
enable.auto.commit: true
enable.auto.offset.store: false
enable.partition.eof: false
linger.ms: 100
enable.idempotence: true
```


@davidblewett We can look into handling explicitly the errors you speak about, but we basically construct new consumer on any error.

```rust
 loop {
        let consumer = Arc::new(try_create_consumer_from_config(
            kafka_config.consumer.clone(),
        )?);
        ...
        consumer.subscribe(&topics)?;

        match run(
            ... (consumer.stream ... inside -> anyhow::Result<()>)
        )
        .await
        {
            Ok(()) => break Ok(()),
            Err(error) => warn!(?error, "Processor failed"),
        }
    }
```

**Comment by zhrebicek** [comment]
One addition we had no issues for producers (`FuruteProducer`), and we still have it in couple of services running for couple of days. So no need to test that @scanterog

**Comment by scanterog** [comment]
I'm surprised we're seeing a working case of the producer with @zhrebicek and a non working one with @neoeinstein. I wonder what the differences would be here.

**Comment by zhrebicek** [comment]
@scanterog We do it like this
```Rust
consumer
        .stream()
        .take_until(token.wait_for_shutdown())
        .err_into::<anyhow::Error>()
        ... rest of pipe working with messages
                .try_buffered(converter_config.max_in_flight)
        .try_for_each(|message| {
            let span = message.span();
            let consumer = consumer.clone();
            async move {
                consumer.store_offset(message.topic(), message.partition(), message.offset())?;
                Ok(())
            }.instrument(span)
        })
        .await

**Comment by scanterog** [comment]
@zhrebicek yeah that translates to stopping the stream entirely when an error is encountered. If a kafka cluster is under high load, then most likely you'll have a lot of requests retries that will be bubbled up as a transient errors. That's usually persistent which translates to the consumer not making progress at all and you'll be recreating the consumers constantly. The only thing that does not add up to me here is that I'd expect still the error to be logged everytime the stream stop and the consumer is recreated. Are you logging these events?

One potential solution is to stop forwarding those transient errors to the app (as with the callback API) but sadly rdkafka does not provide any api to identify those and will require to scan the rdkafka code base and filter those codes in our side.

**Comment by zhrebicek** [comment]
@scanterog we are logging all the errors, and there are no consequent errors after the thing is stuck, so it should not be recreating in loop.

**Comment by scanterog** [comment]
@zhrebicek I see. We haven't able to reproduce this so far. We have tests running for over 5 days and we have tried several operations on the cluster (rolling restarts, broker replacements, ungraceful shutdown). I'll try to setup a service using the pattern you've shared before and see whether we can get closer to understand what's going on.

**Comment by zhrebicek** [comment]
@scanterog If I would be able to reproduce on `staging` I would deploy the version on purpose and enable `rdkafka` debug logs, but it never happened there. And it caused issues in production.

**Comment by zhrebicek** [comment]
@scanterog 
Version: `confluent 6.1.0 (kafka 2.7.0 equivalent)`
Replication factor `2`, in one env it was by mistake 4 and it happened there a bit more then in other envs.

**Comment by scanterog** [comment]
@zhrebicek would it be possible to run a shadow service (that send all reads to dev/null) with this version and debug level enabled (same for rdkafka logs) on the environment where you're getting this issue? Sadly we aren't able to reproduce this issue so there must be something specific we're missing.

@neoeinstein this also applies for the FutureProducer. If you can provide that information would be extremely useful for the project!

Thanks both!

**Comment by zhrebicek** [comment]
I will see what time allows me to try.

**Comment by scanterog** [comment]
Thanks @zhrebicek ! appreciate that

**Comment by Ten0** [comment]
Same problem here

**Comment by davidblewett** [comment]
@Ten0 until we have a way of reproducing the issue, we can't work on a fix. That commit does change the underlying implementation, but aside from some errors bubbling up that didn't before there shouldn't have been a change in behavior. It is also the same mechanism that the Confluent Kafka libraries use. This change has been in use and seeing a lot of traffic and we haven't observed the reported issue.

Are you sure you aren't accidentally swallowing some exceptions? We need a self contained example that triggers the behavior to investigate further.

**Comment by Ten0** [comment]
> Are you sure you aren't accidentally swallowing some exceptions?

Yes I am very sure that we aren't accidentally dropping any error either returned by this library or printed to `log`.
There is no message besides the ones related to too many open files, and those come from the dedicated threads without going through log.

> there shouldn't have been a change in behavior. This change has been in use and seeing a lot of traffic and we haven't observed the reported issue.

From what I read in this topic there have been 3 different people encountering this precise issue (same blockage, same memory increase...) since this precise release: @zhrebicek @neoeinstein and myself. I strongly believe it is time to acknowledge that this issue indeed does exists and has a strong impact despite a minimal reproduction not being established yet.

> Would you please consider yanking the affected releases (and maybe have them only as release candidates) until this is fixed?
(Also this deserves the bug tag)

**Comment by DoumanAsh** [comment]
I'm not sure it is related, but upgrading to 0.36.0 makes topic deletion hang forever, if you have uncommitted messages
JFYI

**Comment by pi-xel** [comment]
I think I was able to reproduce the issue locally. Running high-level consumer with following settings:

```
max.poll.interval.ms=20000
session.timeout.ms=10000
auto.commit.interval.ms=5000
```

After calling `subscribe` I am polling the consumer in a loop using `recv`. If I add a `tokio::time::sleep` of 7s between polls, 0.36.2 silently fails & just stops consuming, with a `sleep` of 6s it resumes. With 0.35.0 it keeps consuming, even using larger `sleep` durations (10s+).

**Comment by pi-xel** [comment]
Minimal example to reproduce the issue (included in the crate's high-level consumer integration tests):

```
// All produced messages should be consumed.
 #[tokio::test(flavor = "multi_thread")]
 async fn test_produce_consume_delay() {
     let _r = env_logger::try_init();
 
     let start_time = current_time_millis();
     let topic_name = rand_test_topic("test_produce_consume_delay");
     let message_map = populate_topic(&topic_name, 10, &value_fn, &key_fn, None, None).await;
     let consumer = create_stream_consumer(&rand_test_group(), None);
     consumer.subscribe(&[topic_name.as_str()]).unwrap();
 
     for i in 0..10 {
         println!("awaiting message: {}", i);
 
         let message = consumer.recv().await;
 
         let timeout = tokio::time::sleep(std::time::Duration::from_millis(5000));
 
         println!("processing message: {}", i);
 
         match message {
             Ok(m) => {
                 let id = message_map[&(m.partition(), m.offset())];
                 match m.timestamp() {
                     Timestamp::CreateTime(timestamp) => assert!(timestamp >= start_time),
                     _ => panic!("Expected createtime for message timestamp"),
                 };
                 assert_eq!(m.payload_view::<str>().unwrap().unwrap(), value_fn(id));
                 assert_eq!(m.key_view::<str>().unwrap().unwrap(), key_fn(id));
                 assert_eq!(m.topic(), topic_name.as_str());
             }
             Err(e) => panic!("Error receiving message: {:?}", e),
         };
 
         timeout.await;
     }
 }
```

On v0.35.0 this completes w/o issues, while on v0.36.2 it gets stuck after the first message w/o any error.

**Comment by pi-xel** [comment]
Interestingly, the issue seems to be directly related to `statistics.interval.ms`:

If I bump it from 500 to 5000ms, the consumer keeps working for any poll delays at or below 5000ms and fails for values above.

**Comment by jwilm** [comment]
I just ran into this issue; might I suggest yanking all variants of the 0.36 branch until a proper patch lands? This will prevent others from unknowingly having issues.

**Comment by untitaker** [comment]
I don't think that's how yank is supposed to be used and I am sure it will cause even more problems for existing users. one example is libraries that depend on 0.36.

**Comment by scotow** [comment]
> I don't think that's how yank is supposed to be used and I am sure it will cause even more problems for existing users. one example is libraries that depend on 0.36.

Libraries that depend on 0.36 might get a warning. But people starting a new project or upgrading their dependencies stacks might introduce this new bug in their software.


As said above, I've noticed similar behaviour using 0.36.2 in two separate projects that use a `StreamConsumer`. Getting some memory leaks after a long period of time (first leak started after 10 days of runtime).

Feel free to tell me if you want some information about my setup (configuration, topics setup, ...).

![image](https://github.com/fede1024/rust-rdkafka/assets/7684550/530696b7-fcb2-4c58-b1a6-b1405a1cabef)

Will rollback to 0.35.0 to see if it fixes the issue.

**Comment by untitaker** [comment]
if i want to depend on a library that depends on 0.36, I cannot do that anymore, and some (if not most) libraries' CI does not check in their lockfiles. So I do think yanking will 100% break somebody's build when it doesn't have to. If you want to fix the onboarding experience you could re-publish 0.34 as 0.37.

**Comment by untitaker** [comment]
> then that library is forcing you to use a version of rdkafka that I wouldn't consider stable for production usage and is functionally broken for some usage pattern

I think that's the point of contention. We use 0.36 just fine and the maintainers have trouble reproducing the issue. It's not universally broken.

**Comment by jwilm** [comment]
If yanking is considered to be too problematic, then perhaps reverting the event API work and publishing a patch version of 0.36 would be more appropriate. Several users have reported production issues with the current release, and it would be good stewardship to prevent others from dealing with that headache.

**Comment by Ten0** [comment]
There seems to be a minimal repro above as well now so maybe fixing it could also be an option.

**Comment by cpaika** [comment]
We're seeing the same problem with `0.36.2` - high level stream consumer stops consuming, nothing in the logs indicating the cause of the hang - CPU drops to 0, and memory slowly increases until the application get's OOMKilled by our Kubernetes cluster.

**Comment by davidblewett** [comment]
Now that this landed, can you test against the latest `master` to see if you still encounter the behavior? https://github.com/fede1024/rust-rdkafka/pull/666 .

**Comment by untitaker** [comment]
there are many other bugfixes in master, it seems a release is warranted anyway

**Comment by svanharmelen** [comment]
Is it confirmed that #666 fixes this issue, or is that still an open question?

**Comment by Ten0** [comment]
It's an open question specifically since nobody confirmed here that it solves the issue.

**Comment by svanharmelen** [comment]
I can compile some of our services against main and let them run in our acceptance env for a while, but we were unable to properly reproduce the issue on demand. It just seem to happen every now and then.

So not sure how representative that test would be...

**Comment by svanharmelen** [comment]
Just read the repro suggestion a couple of comments ago, can play with that as well I guess.

**Comment by svanharmelen** [comment]
As expected reproducing it in our stack proved to be quite difficult. But when running the integration test provided by @pi-xel I can confirm that this test fails when using version 0.36.2 and succeeds when using either 0.35.0 or master. So to the best of my knowledge I would say this issue is resolved using #666 and can be closed.

Since there are more fixes in master (especially #686 which we now hit consistently), it would be great if we could get a new release 🙏🏻

**Comment by davidblewett** [comment]
I released `rdkafka=v0.36.3` and `rdkafka-sys=v4.8.0+2.3.0` today. Based on the length of this thread and the success mentioned earlier, I'm going to close this issue. If you encounter issues with the new release, please file new issues.
